### ISSUE TITLE
Detection of goroutine leaks

### ISSUE THREAD

**Issue by indradhanush** [sourcegraph/sourcegraph#21027]
The inspiration for this came out of a discussion on slack about debugging a customer issue where we had a suggestion about if we were leaking go routines. 

At the moment it appears that we don't have a way of detecting this and I did a quick search for available tools and this showed up: https://github.com/uber-go/goleak.

I wanted to give it a quick run without spending too much time initially to see what it's worth. I chose the `gitserver` and `repoupdater` packages because they are one of the core functions of our product here: https://github.com/sourcegraph/sourcegraph/compare/ig/goleak and surely enough the [build failed](https://buildkite.com/sourcegraph/sourcegraph/builds/95087#ca726a5b-64e3-4977-920b-44e009d1bb0d/92-331). 

The build doesn't indicate very precisely the leaking goroutine but offers some indication. Narrowing it down and eventually fixing it is a slightly more involved process as documented in the readme of [goleak](https://github.com/uber-go/goleak#determine-source-of-package-leaks).

I am creating this issue to share this with the rest of the engineering team. The next steps for this would be:

1. Create a PoC pull request where we use this tool to find goroutine leaks in our code base.
2. Document the steps and create an RFC to start an initiative across all engineering teams to fix goroutine leaks and proactively catch them in our build pipeline. 

I estimate this to be around 1-2 days of work in all and intend to pick it up when I have some free cycles in the upcoming iterations.

In the meantime, if anyone has had experience with similar work and / or wants to pick this up, feel free to assign this issue to yourself or share your ideas and thoughts in this issue. :smile:

**Comment by camdencheek** [comment]
+1 for goleak. I've used it in the past, and it was a very nice sanity check.

**Comment by unknwon** [comment]
> In the meantime, if anyone has had experience with similar work and / or wants to pick this up, feel free to assign this issue to yourself or share your ideas and thoughts in this issue. 😄

Not kidding, I would just search for `go` keyword and audit every use case of it, see if there is an "exit" condition.

I'm not saying goleak is completely useless but I suspect its usefulness without running it continuously in production.

**Comment by tsenart** [comment]
> I'm not saying goleak is completely useless but I suspect its usefulness without running it continuously in production.

Yeah, I agree, I think this would need to be complemented with default monitoring of all our go services for go-routine leaks, which should be relatively straightforward. Without the monitoring of trends in number of go-routines, I'm afraid go-leak could induce a false sense of security.

**Comment by slimsag** [comment]
I have no context on what is being discussed here, just giving a drive-by-comment that our Prometheus metrics include # of goroutines per service in case leveraging that info is useful at all to you. [Example here](https://sourcegraph.com/-/debug/grafana/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22go_goroutines%22,%22requestId%22:%22Q-91fd5f6e-67bc-469a-a1b2-65142f1c0189-0A%22%7D%5D) (we don't expose this in any of the dashboards we have currently, AFAIK)
